[NEW ZOO-FUNCTIONS] Requisito 4 - isManager (#22)
betrybe/sd-0x-project-zoo-functions: README.md

## Sumário

- [Boas vindas ao repositório do projeto de ES6 e Higher Order Functions!](#boas-vindas-ao-repositório-do-projeto-de-es6-e-higher-order-functions)
  - [Habilidades](#habilidades)
  - [Sumário](#sumário)
  - [Entregáveis](#entregáveis)

>     - [O QUE DEVERÁ SER DESENVOLVIDO](#o-que-deverá-ser-desenvolvido)

Reviewer: ```suggestion
    - [O que deverá ser desenvolvido](#o-que-deverá-ser-desenvolvido)
```

    - [DESENVOLVIMENTO](#desenvolvimento)
    - [DATA DE ENTREGA](#data-de-entrega)
  - [Instruções para entregar seu projeto](#instruções-para-entregar-seu-projeto)
    - [ANTES DE COMEÇAR A DESENVOLVER](#antes-de-começar-a-desenvolver)
    - [DURANTE O DESENVOLVIMENTO](#durante-o-desenvolvimento)
    - [DEPOIS DE TERMINAR O DESENVOLVIMENTO (OPCIONAL)](#depois-de-terminar-o-desenvolvimento-opcional)
  - [Requisitos do projeto](#requisitos-do-projeto)
    - [1. IMPLEMENTE A FUNÇÃO getSpeciesByIds](#1-implemente-a-função-getspeciesbyids)
    - [2. IMPLEMENTE A FUNÇÃO getAnimalsOlderThan](#2-implemente-a-função-getanimalsolderthan)
    - [3. IMPLEMENTE A FUNÇÃO getEmployeeByName](#3-implemente-a-função-getemployeebyname)

>     - [4. IMPLEMENTE AS FUNÇÕES isManager E employeeManagement](#4-implemente-as-funções-ismanager-e-employeemanagement)

Reviewer: ```suggestion
    - [4. Implemente as funções `isManager` e `employeeManagement`](#4-implemente-as-funções-ismanager-e-employeemanagement)
```

    - [5. IMPLEMENTE A FUNÇÃO countAnimals](#5-implemente-a-função-countanimals)
    - [6. IMPLEMENTE A FUNÇÃO calculateEntry](#6-implemente-a-função-calculateentry)
    - [7. IMPLEMENTE A FUNÇÃO getAnimalMap](#7-implemente-a-função-getanimalmap)
    - [8. IMPLEMENTE A FUNÇÃO getSchedule](#8-implemente-a-função-getschedule)
    - [9. IMPLEMENTE A FUNÇÃO getOldestFromFirstSpecies](#9-implemente-a-função-getoldestfromfirstspecies)
    - [10. IMPLEMENTE A FUNÇÃO getEmployeeCoverage](#10-implemente-a-função-getemployeecoverage)
  - [Avisos Finais](#avisos-finais)
    - [REVISANDO UM PULL REQUEST](#revisando-um-pull-request)
    - [AVALIAÇÃO DO PROJETO](#avaliação-do-projeto)

>     - [ORDEM DOS REQUISITOS NO AVALIADOR AUTOMÁTICO](#ordem-dos-requisitos-no-avaliador-automático)

Author: Oi @cleytonoliveira vc quer dizer em relação a que especificamente?

>     - [ORDEM DOS REQUISITOS NO AVALIADOR AUTOMÁTICO](#ordem-dos-requisitos-no-avaliador-automático)

Reviewer: Aos títulos. Acredito que seja pra evitar caixa alta. Acredito que fez isso por que esta em caixa alta ao decorrer do Readme, mas seria ideal corrigir eles. Vou deixar como sugestão.

  6. `npm test` _(executa todos os testes presentes na aplicação)_
  7. `npm test path/to/file` _(executa apenas os testes presentes no arquivo path/to/file)_

#### DICA PARA O DESENVOLVIMENTO

> Em sua vida de pessoa desenvolvedora, saber quebrar problemas maiores em menores ajudará a resolver problemas mais complexos de maneira mais fácil. Além disso, funções são blocos de códigos lógicos com uma função específica. Dividir suas funções em pequenos blocos de lógica é bem útil, torna tudo mais fácil de entender e você sempre pode reutilizar esse bloco de código em outras partes de sua aplicação.

Reviewer 2: Uma sugestão para esse trecho: 

`Em sua vida de pessoa desenvolvedora, o pensamento algoritmico, processo de quebrar problemas complexos em pequenas partes para resolvê-los, ajudará a solucionar de maneira mais fácil os problemas complexos. Assim, uma boa forma de começar a desenvolver um projeto é lembrar que as funções são compostas por pequenos blocos de lógica que têm finalidade (função) específica. Ou seja, quebrar as funções em pequenos blocos de lógica tornará tudo mais simples de se entender e você sempre poderá reutilizar esses códigos em outras partes de sua aplicação. `

Acho que ciatr a questão do pensamento algoritmico pode ajudar a apsaar a ideia de a pessoa dividir os problemas em pequenas partes para resolver.

Antes de começar, analise o arquivo `src/data.js`, para ver os dados que serão usados.

> ### 1. Implemente a função getSpeciesByIds

Reviewer 3: O que acha?

```suggestion
### 1. Implemente a função `getSpeciesByIds`
```

E se fizer sentido, aplicar nos outros locais

  - Quando provido o primeiro nome do funcionário, retorna o objeto do funcionário
  - Quando provido o último nome do funcionário, retorna o objeto do funcionário

> ### 4. IMPLEMENTE AS FUNÇÕES isManager E employeeManagement

Reviewer: ```suggestion
### 4. Implemente as funções `isManager` e `employeeManagement`
```

> Em sua vida de pessoa desenvolvedora, saber quebrar problemas maiores em menores ajudará a resolver problemas mais complexos de maneira mais fácil. Além disso, funções são blocos de códigos lógicos com uma função específica. Quebrar suas funções em pequenos blocos de lógica é bem útil, torna tudo mais fácil de entender e você sempre pode reutilizar esse bloco de código em outras partes de sua aplicação.

Reviewer: Acha que faz sentido esse paragrafo só para este requisito? Acredito que terá mais requisitos que vão utilizar deste padrão, não faria sentido colocar na parte de desenvolvimento ou algo assim?

Reviewer 4: Ficou bem legal ter essa orientação para o pessoal! Porém concordo com o Cleyton, como teremos outros requisitos com essa prática creio que o ideal seria colocar em uma seção de dicas e não dentro do requisito. Seria algo como:

Dicas:
-  Quebrar suas funções em pequenos blocos de lógica tem inúmeras vantagens: reduz a complexidade, facilita a detecção de erros e a manutenção do código, além de que você poderá reutilizar esse bloco de código em outras partes de sua aplicação. Você verá nesse projeto que algumas funções se utilizam de outras funções de forma a se complementarem, facilitando a resolução. Utilize essa prática quando considerar útil para resolver problemas mais complexos!

> Pensando em tudo isso, dividimos para você a função employeeManagement em duas funções. A primeira, isManager, é responsável por verificar se um `id` passado por parâmetro é de uma pessoa colaboradora gerente ou não. A segunda função, employeeManagement utiliza da informação da primeira função da seguinte maneira:

Reviewer: ```suggestion
Pensando em tudo isso, dividimos para você a função `employeeManagement` em duas funções. A primeira, `isManager`, deverá ser responsável por verificar se um `id` passado por parâmetro é de uma pessoa colaboradora gerente ou não. A segunda função, `employeeManagement` deverá utilizar a informação da primeira função considerando as seguintes condições:
```

Reviewer 4: Aqui, relacionando com o parágrafo sobre a prática de dividir em blocos de código menores, uma sugestão seria:

Considerando a boa prática de dividir o código em partes menores, apresentamos a função employeeManagement em que você deverá dividí-la em duas funções: a primeira - isManager - que será responsável por verificar uma pessoa colaboradora é gerente ou não, e a segunda - employeeManagement - que utiliza a primeira função para apresentar as seguintes saídas:

* Se ocupar cargo de gerência, a sua função deve retornar um array com os nomes das pessoas colaboradoras que ela gerencia. 
* Se não ocupar cargo de gerência, a sua função deve retornar um array contendo os id's de seus gerentes.

  **Observações técnicas**

> - Retorna um array contendo nome e sobrenome das pessoas colaboradoras gerenciadas por determinada pessoa com cargo de gerência.
>     Exemplo de output: // [ 'Burl Bethea', 'Ola Orloff', 'Emery Elser' ];

Reviewer: ```suggestion
- Retorna um array contendo nome e sobrenome das pessoas colaboradoras gerenciadas por determinada pessoa com cargo de gerência.
    Exemplo de output:
```javascript
[ 'Burl Bethea', 'Ola Orloff', 'Emery Elser' ];
```
```

- Retorna um array contendo os id's dos gerentes que são responsáveis pela pessoa colaboradora.

>     Exemplo de output: // [ '0e7b460e-acf4-4e17-bcb3-ee472265db83', 'fdb2543b-5662-46a7-badc-93d960fdc0a8' ];

Reviewer: ```suggestion
    Exemplo de output:
```javascript
[ '0e7b460e-acf4-4e17-bcb3-ee472265db83', 'fdb2543b-5662-46a7-badc-93d960fdc0a8' ];
```
```

  **O que será avaliado**

  - Se o id passado for de um gerente, retorna um array contendo nome e sobrenome das pessoas colaboradoras que ela é responsável;

>   - Se o id passado **não** for de um gerente, retorna um array contendo os id's dos gerentes responsáveis por ela;

Reviewer: ```suggestion
  - Se o id passado **não** for de um gerente, retorna um array contendo os id's das pessoas gerentes responsáveis por ela;
```

  - Quando provido o primeiro nome do funcionário, retorna o objeto do funcionário
  - Quando provido o último nome do funcionário, retorna o objeto do funcionário

### 4. Implemente as funções isManager E employeeManagement

>   Considerando a boa prática de dividir o código em partes menores, apresentamos a função employeeManagement em que você deverá dividí-la em duas funções: a primeira - isManager - que será responsável por verificar se uma pessoa colaboradora é gerente ou não, e a segunda - employeeManagement - que utiliza a primeira função para apresentar as seguintes saídas:

Reviewer 5: Ale, eu achei um pouco confusa a descrição do exercício.
Pra ficar mais claro poderia ser algo como 
```suggestion
Considerando a boa prática de dividir o código em partes menores, neste requisito você deve dividir o seu código em duas funções: 
1. isManager - que será responsável por verificar se uma pessoa colaboradora é gerente ou não 
2. employeeManagement - que utiliza a primeira função para apresentar as seguintes saídas:
```

---

### 4. Implemente a função `getRelatedEmployees`

Considerando a boa prática de dividir o código em partes menores, apresentamos a função `getRelatedEmployees` em que você deverá dividí-la em duas funções: 
  
  1 - `isManager` - que será responsável por verificar se uma pessoa colaboradora é gerente ou não. O retorno dessa função deve ser um booleano: `true`ou `false`;
  2 - `getRelatedEmployees` - que utiliza a primeira função para apresentar as seguintes saídas: 
     * se for uma pessoa colaboradora gerente, deve retornar um array contendo os nomes das pessoas colaboradoras que ela é responsável;
     * se **não** for uma pessoa colaboradora gerente, deverá ser lançado um error gerado com a função construtora **Error** da biblioteca padrão do JavaScript com a mensagem **"O id inserido não é de uma pessoa colaboradora gerente!"**. Exemplo:

```javascript

> throw new Error('Informações inválidas');

Reviewer 6: Será que uma dessas opções é válida?

- Ou uma sugestão de pesquisa: "Lançar objeto de erro javascript"
- Ou o link da [documentação](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Error)

> ### 4. Implemente a função `isManager`

Reviewer: ```suggestion
### 4. Implemente a função `getRelatedEmployees`
```
O que você acha? Acredito que devemos considerar a função que tem a saída final. Faz sentido?

Considerando a boa prática de dividir o código em partes menores, apresentamos a função `getRelatedEmployees` em que você deverá dividí-la em duas funções: 
  
  1 - `isManager` - que será responsável por verificar se uma pessoa colaboradora é gerente ou não, e a segunda;

>   2 - `getRelatedEmployees` - que utiliza a primeira função para apresentar as seguintes saídas: 

Reviewer: Quais saídas? O próximo tópico é observações técnicas, acho que da pra deixar mais nítido, o que acha?

Author: Nossa passou batido essa parte mesmo hahaha

> Considerando a boa prática de dividir o código em partes menores, apresentamos a função `getRelatedEmployees` em que você deverá dividí-la em duas funções: 
>   
>   1 - `isManager` - que será responsável por verificar se uma pessoa colaboradora é gerente ou não, e a segunda;
>   2 - `getRelatedEmployees` - que utiliza a primeira função para apresentar as seguintes saídas: 

Author: ```suggestion
Considerando a boa prática de dividir o código em partes menores, apresentamos a função `getRelatedEmployees` em que você deverá dividí-la em duas funções: 
  
  1 - `isManager` - que será responsável por verificar se uma pessoa colaboradora é gerente ou não. O retorno dessa função deve ser um booleano: `true`ou `false`;
  2 - `getRelatedEmployees` - que utiliza a primeira função para apresentar as seguintes saídas: 
     * se a pessoa colaboradora tiver cargo de gerente, deve retornar um array contendo os nomes das pessoas colaboradoras que ela é responsável;
     * se a pessoa colaboradora não tiver cargo de gerente, deve disparar um erro.
  
  
```

> 

Reviewer: Senti falta de informação sobre como lançar o erro, até porque este requisito esta antes do que o Paulo fez. Veja o que da pra adaptar do que o Paulo fez.
